tests/gold_tests/redirect/redirect-deny-loopback.test.py

```diff
@@ -0,0 +1,143 @@
+            'proxy.config.diags.debug.enabled': 1,
+            'proxy.config.diags.debug.tags': 'http|dns|redirect',
+            'proxy.config.http.number_of_redirections': 1,
+            'proxy.config.http.cache.http': 0,
+            'proxy.config.dns.nameservers': '127.0.0.1:{0}'.format(dns.Variables.Port),
+            'proxy.config.dns.resolv_conf': 'NULL',
+            'proxy.config.url_remap.remap_required': 0,
+            'proxy.config.http.redirect_loopback_mode': scenario.value['config'],
+            'proxy.config.http.connect_attempts_timeout': 5,
+            'proxy.config.http.connect_attempts_max_retries': 0,
+        })
+        tr.Processes.Default.StartBefore(trafficservers[scenario])
+    else:
+        tr.StillRunningAfter = trafficservers[scenario]
+
+    testDomain = 'testdomain{0}.test'.format(redirectTarget.replace('.', 'x'))
+    # The micro DNS server can't tell us whether it has a record of the domain already, so we use a dictionary to avoid duplicates.
+    dnsRecords[testDomain] = [redirectTarget]
+
+    # A GET request parameterized on the scenario and on the target.
+    request_header = {
+            'headers': ('GET /redirect?scenario={0}&target={1} HTTP/1.1\r\n'
+                        'Host: *\r\n\r\n').\
+                                format(scenario.name, redirectTarget),
+            'timestamp': ArbitraryTimestamp,
+            'body': ''}
+    # Returns a redirect to the test domain for the given target & the port number for the TS of the given scenario.
+    response_header = {
+            'headers': ('HTTP/1.1 307 Temporary Redirect\r\n'
+                        'Location: http://{0}:{1}/\r\n'
+                        'Connection: close\r\n\r\n').\
+                                format(testDomain, origin.Variables.Port),
+            'timestamp': ArbitraryTimestamp,
+            'body': ''}
+    origin.addResponse('sessionfile.log', request_header, response_header)
+
+    # Generate the request data file.
+    with open(os.path.join(data_path, tr.Name), 'w') as f:
+        f.write(('GET /redirect?scenario={0}&target={1} HTTP/1.1\r\n'
+                 'Host: iwillredirect.test:{2}\r\n\r\n').\
+                         format(scenario.name, redirectTarget, origin.Variables.Port))
+    # Set the command with the appropriate URL.
+    tr.Processes.Default.Command = 'python tcp_client.py 127.0.0.1 {0} {1} | head -n 1'.\
+            format(trafficservers[scenario].Variables.port, os.path.join(data_dirname, tr.Name))
+    # Generate and set the 'gold file' to check stdout
+    goldFilePath = os.path.join(data_path, '{0}.gold'.format(tr.Name))
+    with open(goldFilePath, 'w') as f:
+        f.write(scenario.value['expectedStatusLine'])
+    tr.Processes.Default.Streams.stdout = goldFilePath
+
+# This test does not detect available loopback addresses.
+# We can add them here for manual testing.
+for t in (['127.0.0.1']):
+    for s in ScenarioE:
+        makeTestCase(redirectTarget=t, scenario=s)
```

Comment:
For automated tests that gate things like PR acceptance, this should be sensible.

However, a more complete test would include a collection of different ways to address a loopback device. This is difficult to accommodate at the level where the test runs, as we would need infrastructure in the loop to arrange for interfaces at such addresses.

So that is why I have left the outer loop here with comments to the effect.

Reply:
This is now picking all the non-link-local IPv6 as well as IPv4 addresses of the host.